sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DefaultColumns.scala

```diff
@@ -0,0 +1,287 @@
+        if table.rows.nonEmpty && table.rows.forall(_.size == table.rows(0).size) =>
+        table.rows(0).size
+      case project: Project => project.projectList.size
+      case _ => return insert
+    }
+    // The table value provides the DEFAULT column values as text; analyze them into expressions.
+    val schema: StructType = getInsertTableSchema(insert, catalog).getOrElse(return insert)
+    val schemaWithoutPartitionCols = StructType(schema.fields.dropRight(insert.partitionSpec.size))
+    val coerced: Seq[Expression] = for {
+      field <- schemaWithoutPartitionCols.fields.drop(numQueryOutputs)
+      name: String = field.name
+      text: String =
+        if (field.metadata.contains(default)) field.metadata.getString(default) else "NULL"
+      // Parse the DEFAULT column expression. If the parsing fails, throw an error to the user.
+      expr: Expression = Parse(name, text, "INSERT")
+      // Analyze and constant-fold each result.
+      analyzed: Expression = Analyze(expr, text, name, "INSERT")
+      // Perform implicit coercion from the provided expression type to the required column type.
+      errorPrefix = "Failed to execute INSERT command because the destination table column "
+      coerced: Expression =
+        if (field.dataType == analyzed.dataType) {
+          analyzed
+        } else if (Cast.canUpCast(analyzed.dataType, field.dataType)) {
+          Cast(analyzed, field.dataType)
+        } else {
+          throw new AnalysisException(errorPrefix +
+            s"$name has a DEFAULT value with type ${field.dataType}, but the " +
+            s"query provided a value of incompatible type ${analyzed.dataType}")
+        }
+    } yield coerced
+    // Finally, return a projection of the original `insert.query` output attributes plus new
+    // aliases over the DEFAULT column values.
+    // If the insertQuery is an existing Project, flatten them together.
+    val newQuery = insert.query match {
+      case Project(projectList, child) =>
+        val newAliases: Seq[NamedExpression] = coerced.zip(schemaWithoutPartitionCols.fields).map {
+          case (expr, field) => Alias(expr, field.name)() }
+        Project(projectList ++ newAliases, child)
+      case table: UnresolvedInlineTable =>
+        val newNames: Seq[String] =
+          schemaWithoutPartitionCols.fields.drop(numQueryOutputs).map { _.name }
+        table.copy(names = table.names ++ newNames, rows = table.rows.map { row => row ++ coerced })
+      case _ => insert.query
+    }
+    insert.copy(query = newQuery)
+  }
+
+  /**
+   * Replaces unresolved "DEFAULT" column references with matching default column values.
+   *
+   * Background: CREATE TABLE and ALTER TABLE invocations support setting column default values for
+   * later operations. Following INSERT, and INSERT MERGE commands may then reference the value
+   * using the DEFAULT keyword as needed.
+   *
+   * Example:
+   * CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
```

Comment:
Hi @srielau, I implemented it this way because the design doc appears to specify that the DEFAULT expression must be folded to a constant [1]. Is this intended? If not, that is OK, I can update this PR to skip the constant-folding and simply validate and store the original text in the column metadata instead.

[1] 
<img width="1316" alt="image" src="https://user-images.githubusercontent.com/99207096/158637309-38338720-2d13-4e32-9df1-8d59e2518f5a.png">

Reply:
Resolved this confusion offline with Serge. This should be evaluated at the INSERT INTO time:
```
Catalyst
The catalyst design is straight forward. The defaults are read in with the rest of the column info.
We leave DEFAULT keywords at DML as unresolved until analysis.
```
I can update the PR accordingly.

Reply:
This is done, the "current default" value obtained from the initial CREATE/REPLACE TABLE statement is now persisted directly without constant-folding. Instead, a new "exists default" column metadata now holds the constant-folded result separately. Following INSERT/UPDATE/MERGE statements can use the former while table scans can use the latter.